### ISSUE TITLE
Add Visual Studio Code PowerShell extension formatting settings to DSC Resource template

### ISSUE THREAD

**Issue by PlagueHO** [PowerShell/DscResources#284]
The Visual Studio Code PowerShell Extension contains many automatic formatting settings.

The default settings for these however will format PowerShell code in a way that conflicts with the StyleGuidelines. There may also be other settings in VS Code that can cause style violations and if anyone knows of them please feel free to make suggestions.

Therefore I propose:
1. Include a `.vscode\settings.json` file in the DSCResource Template containing the settings that are correct for DSC Resource Kit resources.
2. Start adding a `.vscode\settings.json` file to each of the DSC Resource kit resources.
3. Update `GettingStartedWithGitHub.md` to change the .gitignore recommendation to allow adding of .vscode\settings.json to tracked.

What I'd like to know is:
1. Are other repos already doing this and I'm just late to the party?
2. Is worth doing? My 2c is that it is.
3. If so, what are the settings that should be defined? Only settings that we actually have style guidelines should be defined IMHO.

I think these are the settings that should be used:
```
    "powershell.codeFormatting.openBraceOnSameLine": false,
    "powershell.codeFormatting.newLineAfterOpenBrace": false,
    "powershell.codeFormatting.newLineAfterCloseBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenParen": true,
    "powershell.codeFormatting.whitespaceAroundOperator": true,
    "powershell.codeFormatting.whitespaceAfterSeparator": true,
    "powershell.codeFormatting.ignoreOneLineBlock": false,
    "powershell.codeFormatting.alignPropertyValuePairs": true
```

And for reference, here are all the settings defined by default:
```
  // Places open brace on the same line as its associated statement.
  "powershell.codeFormatting.openBraceOnSameLine": true,

  // Adds a newline (line break) after an open brace.
  "powershell.codeFormatting.newLineAfterOpenBrace": true,

  // Adds a newline (line break) after a closing brace.
  "powershell.codeFormatting.newLineAfterCloseBrace": true,

  // Adds a space between a keyword and its associated scriptblock expression.
  "powershell.codeFormatting.whitespaceBeforeOpenBrace": true,

  // Adds a space between a keyword (if, elseif, while, switch, etc) and its associated conditional expression.
  "powershell.codeFormatting.whitespaceBeforeOpenParen": true,

  // Adds spaces before and after an operator ('=', '+', '-', etc.).
  "powershell.codeFormatting.whitespaceAroundOperator": true,

  // Adds a space after a separator (',' and ';').
  "powershell.codeFormatting.whitespaceAfterSeparator": true,

  // Does not reformat one-line code blocks, such as "if (...) {...} else {...}".
  "powershell.codeFormatting.ignoreOneLineBlock": true,

  // Align assignment statements in a hashtable or a DSC Configuration.
  "powershell.codeFormatting.alignPropertyValuePairs": true,

  // Shows the integrated console when the PowerShell extension is initialized.
  "powershell.integratedConsole.showOnStartup": true,

  // Switches focus to the console when a script selection is run or a script file is debugged. This is an accessibility feature. To disable it, set to false.
  "powershell.integratedConsole.focusConsoleOnExecute": true,
```

Tagging maintainers:
@johlju, @bgelens, @BrianFarnhill, @RichardSiddaway, @raandree, @mbreakey3, @jcwalker, @tysonjhayes, @athaynes, @mhendric

**Comment by ArieHein** [comment]
I'm not sure having a per-IDE setting is a good thing.

What if I use Visual Studio and not Code, and in my VS setting I have static code analysis rules
that might conflict with the StyleGuidelines ?

Youll have to know how each IDE works and have a special dedicated formatting rule to adhere to the DSC StyleGuidelines.

Maybe adding it as a unit test so you test the rule as described by the StyleGuidelines and fail the test appropriately and then suggest additional reading or link to the StyleGuidelines to show the dev, why and how to fix it, but eventually it is up to the Dev to fix their own IDE.

This way you separate the what from the how.

my 0.5cent

**Comment by BrianFarnhill** [comment]
I really like this idea. We already include other VS Code setting in SharePointDsc so this would be a simple inclusion for us that I think would add good value. I know we can't target every IDE but given the popularity of VS Code in the space we are in here I think it's worth doing for sure.

**Comment by RichardSiddaway** [comment]
I'm good with VS code settings being available

**Comment by raandree** [comment]
Sometimes I still use the ISE but I see also the future is with VS Code if it comes to PowerShell. I am supporting the idea.

**Comment by bgelens** [comment]
## On topic: 
I am in favor of including the vs code settings

## Off topic:
Personally I was never really font of the DscResource style guideline. I always write my code in K&R style and when I'm contributing to the Reskit, it's always hard for me to do it in the "correct" style. This is where the vs code settings on a project basis will help me.

Discussing the style guideline future is however topic for another issue? @raandree I believe you where going to create one for it?

**Comment by johlju** [comment]
I'm in favor of this idea of having workspace settings for VS Code. I have long thought about adding, at least, 
"files.trimTrailingWhitespace": true
"files.insertFinalNewLine": true
These will add a new line at the end of the file and trim any white space at the end of rows. 
These will help with the review. Not getting white space changes, and no files will come in without a new line at the end, unless the user using any other editor. 
I all in favor of adding the settings mentioned above, and any other setting which would help contributors. 

I'm all for supporting workspace settings for other editors also if they support that.

**Comment by johlju** [comment]
@raandree I'm all for changing the guideline too. K&R is the way I have coded previously, although after coding so much in the DSC Resource Kit I'm  changing the default style (in the back of my head) to follow the style guideline. 
I have not a problem keeping the style guideline as is, but if the guideline could be made thinner by using VS Code default then all for changing. 
Many contributors are now coding against the style guideline. A lot of code are using the guideline now. Changing would mean a lot of code is not aligned with the guideline. 

If the masses want to use a new style, I'm all for it.

**Comment by PlagueHO** [comment]
Couple of my thoughts on this:
The auto formatting is only actually applied on demand: when you press SHIFT+ALT+F or select from the F1 method. There is probably a way to have it auto apply - but I'm not proposing that we enable automatically applying style.

So it would be up-to the person making the change to apply the auto formatting - it would just be an easy way to advise contributors to correct style as per guidelines (easier than typing "Please move { to next line" 50 times :grin: ).

The problem is that the default VS Code settings will cause style violations if they are left as default, and I have made this mistake myself and have had a few PR submissions where I think this has also happened (which requires lots PR comments to get fixed up).

But definitely agree with @ArieHein - we need some rules that are applied in AppVeyor to check this, which is also something I'm working on (https://github.com/PowerShell/DscResource.Tests/issues/86) but haven't had time to finish this.

@raandree - I'm OK with whatever style guidelines we use as long as they're consistent across repos. But the thing I try to remember about changing style guidelines/standards is that they introduce tech debt - which most repos are swimming in already (no one likes paying down tech debt right? :grin: ) . So I'd suggest changing style guidelines is something that should be considered in conjunction with introduced tech debt cost.

Anyway, the proposed settings would be (including @johlju 's really good ones):
```
    "powershell.codeFormatting.openBraceOnSameLine": false,
    "powershell.codeFormatting.newLineAfterOpenBrace": false,
    "powershell.codeFormatting.newLineAfterCloseBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenParen": true,
    "powershell.codeFormatting.whitespaceAroundOperator": true,
    "powershell.codeFormatting.whitespaceAfterSeparator": true,
    "powershell.codeFormatting.ignoreOneLineBlock": false,
    "powershell.codeFormatting.alignPropertyValuePairs": true,
    "files.trimTrailingWhitespace": true
    "files.insertFinalNewLine": true
```

Is that correct?

**Comment by johlju** [comment]
@ArieHein I think @PlagueHO has an issue, and working on something around that, to create unit tests for testing if the code is not aligned to the guideline.

**Comment by johlju** [comment]
@PlagueHO I agree at least those we should have. Haven't come across any other that I know could benefit us.

**Comment by johlju** [comment]
As @bgelens pointed out in a PR adding these. This is not part of the style guideline. I totally missed this one. We should not have this in the settings set to $true, but rather $false. Or keep it $true and change the guideline. 
```
"powershell.codeFormatting.alignPropertyValuePairs": false
```

**Comment by PlagueHO** [comment]
Totally correct @bgelens and @johlju - here is the corrected list:
```
    "powershell.codeFormatting.openBraceOnSameLine": false,
    "powershell.codeFormatting.newLineAfterOpenBrace": false,
    "powershell.codeFormatting.newLineAfterCloseBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenParen": true,
    "powershell.codeFormatting.whitespaceAroundOperator": true,
    "powershell.codeFormatting.whitespaceAfterSeparator": true,
    "powershell.codeFormatting.ignoreOneLineBlock": false,
    "files.trimTrailingWhitespace": true,
    "files.insertFinalNewline": true
```

I have this enabled for my VS Code user settings because it produces nicely formatted code, but is not required.

**Comment by johlju** [comment]
It seem that alignPropertyValuePairs is default set to true. Should we then set it to false? 

```
// Align assignment statements in a hashtable or a DSC Configuration.
  "powershell.codeFormatting.alignPropertyValuePairs": true,
```

**Comment by PlagueHO** [comment]
@johlju - I reckon that if it is not defined by the guidelines then we should leave it down to the individual user environment. If we enforce it then we're preventing users from using this setting - which I actually use :grin: regardless of the guidelines because it results in much nicer code.

But if the community was happy to enforce a standard on this then that is fine - but I'd suggest this will add a lot of technical debt (my day job I spend a lot of time lecturing teams about technical debt - so it's hard to switch off :cry: ).

**Comment by johlju** [comment]
@PlagueHO I agree. And agree not to add to the technical debt. I think it's good that you point out that! 😄

**Comment by bgelens** [comment]
So since [today](https://github.com/PowerShell/vscode-powershell/releases/tag/v1.4.0), we can reduce the amount of rules a bit :smile:

```
"powershell.codeFormatting.preset": "Allman"
```

**Comment by johlju** [comment]
Yes, saw that David push a new build with that change today :)

**Comment by cwegener** [comment]
I agree with @ArieHein Maintaining settings for each editor individually does not sound like the best idea.
I would prefer an `.editorconfig` file to ensure that contributors can easily use the basic settings from the StyleGuide. And then more specific StyleGuide checks could probably be implemented via CI.

**Comment by ArieHein** [comment]
@cwegener The reason I didnt go for that idea is basically contributors, or lack of to be exact. Youll need multiple people that use a specific editor that have an understanding about inner works of that editor to make the key-value setting per that editor and stay uptodate..it will just cause unnecessary debt eventually.

The only thing you can control is the pester tests that you can run on someone else's pull request or some sort of a free online "code analysis" prior to accepting it.

**Comment by PlagueHO** [comment]
@cwegener , @ArieHein - The purpose of this change is not to prevent people making or submitting style violations. [This issue](https://github.com/PowerShell/DscResource.Tests/issues/86) is to prevent that. I'm working on that.

So with the above statement in mind, what we're trying to address is (using standard user story format):
1.
As a PR reviewer
I want to have to make less style comments for uses who use VS Code and have applied autoformatting
So that I can spend more time working on the PSSA rules to warn/block style violations: https://github.com/PowerShell/DscResource.Tests/issues/86

2.
As a PR contributor
I want an quite and easy way to bring a DSC Resource Kit PS1/PSM1 file into HQRM style 
So that I can rapidly reduce technical debt caused by style violations with the least amount of effort

I also don't see that adding the settings for VS Code means that we have to add them for every IDE - because again, this is not to solve the problem of style violations. It is to **make them easier to fix** and to **prevent users of VS Code from completely messing up a file** by pressing <kbd>SHIFT</kbd>+<kbd>ALT</kbd>+<kbd>F</kbd> without having the correct style guidelines set up.

One final thing: Many projects I work on use much different style guidelines than the ones for the DSC Resource kit - so setting up auto-formatting rules on a per repo basis is hugely handy to me. That way I can work on lots of different projects and don't have to remember which style settings I need to apply. So this is hugely handy for those of us who try to spread ourselves quite thin.

Oh and @bgelens - oh that is cool! I'll update the list.

**Comment by BrianFarnhill** [comment]
@PlagueHO A big plus one to that, I don't believe the intent here is to apply settings for every IDE - its about adding a file that will help a very common IDE for exactly those scenarios. 

But realistically, what other IDEs are people using for PowerShell DSC modules - just asking the question? I know some use the full Visual Studio and as far as I'm aware we couldn't control the settings for that through a file in the source code like we get with VS code. If there are other IDEs out there that we can set these kind of formatting "preferences" for, I'm all ears for it though.

Just my 2c

**Comment by bgelens** [comment]
FWI: When looking how Allman is setup by the [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer/pull/782/files#diff-fb26e91bfb6f12fa4297bf5f17d10ffa) (which handles the formatting in vs code), I'm not convinced the style guidelines are 1:1 with this. Better keep the rules individually.

**Comment by PlagueHO** [comment]
@bgelens - Awesome you did some digging - I meant to do that as well because I wasn't sure which settings we could replace with this. So will keep them as is for now.

**Comment by bgelens** [comment]
I figured out today that the workspace settings are not honored if the user settings contain ```"powershell.codeFormatting.preset"``` configured. So when we point users in the direction on how to trigger the formatting, this is crucial info.

In the meantime I raised an [issue](https://github.com/PowerShell/vscode-powershell/issues/933) at the PowerShell extension repo

**Comment by bgelens** [comment]
So, one to add to our set is ```"powershell.codeFormatting.preset": "Custom"``` this will make sure my previous comment doesn't apply

**Comment by johlju** [comment]
@bgelens Nice catch!

**Comment by PlagueHO** [comment]
So new list is:
```
    "powershell.codeFormatting.openBraceOnSameLine": false,
    "powershell.codeFormatting.newLineAfterOpenBrace": false,
    "powershell.codeFormatting.newLineAfterCloseBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenBrace": true,
    "powershell.codeFormatting.whitespaceBeforeOpenParen": true,
    "powershell.codeFormatting.whitespaceAroundOperator": true,
    "powershell.codeFormatting.whitespaceAfterSeparator": true,
    "powershell.codeFormatting.ignoreOneLineBlock": false,
    "powershell.codeFormatting.preset": "Custom",
    "files.trimTrailingWhitespace": true,
    "files.insertFinalNewline": true
```

**Comment by bgelens** [comment]
@PlagueHO still missing:
```
"powershell.codeFormatting.alignPropertyValuePairs": false
```

**Comment by PlagueHO** [comment]
@bgelens - I'm not sure we should enforce that one to false, because there isn't a style guideline for it.

I set it to True on on my own machine, because I personally like to use it. But I wouldn't raise a comment on someone elses code if it wasn't aligned. Setting this to False will prevent me from using my own setting value- which is why I left it out. Does that make sense?

**Comment by bgelens** [comment]
it makes sense but.. 

your setting is actually the default setting and so will it be (unknowingly) for a lot of contributors. So in the end it would probably (naturally) result in all resources having this alignment. Why not have it in the style guideline and add it here as ```"powershell.codeFormatting.alignPropertyValuePairs": true``` to make it explicit?

I'm fine either way. I just need to know what to put in the modules :smile:

**Comment by PlagueHO** [comment]
@bgelens - Ah, I see what you mean. That is a tricky one. I prefer that style for sure, but wouldn't want to say it should be enforced across all repos. Maybe leave it open for discussion for a bit and see what everyone says? I'll cope either way :grin:

**Comment by ronx** [comment]
been following this thread for a while and just wanted to give my vote for making this setting part of the standard. A couple of months ago, I "standardized" a lot of my own code and borrowed snippets to remove the alignment because I dig consistency and the "standards" didn't seem to support the alignment. However, that's one preference that I had not grown accustomed to either way and was totally unbiased about, yet found myself feeling dissonance in my code without the alignment readability. After working with PS and DSC structures in the wild, I think this should be added to the standards to support readability and maintain sanity. That said, I do recognize the implications for those that use other text editors that may or may not have the niceties that our precious vscode (et al.) have to be flexible to our developer intricacies.

So, meeting in the middle, I'd at least suggest a default of "true" for vscode even if the standard doesn't specify a preference ;)

**Comment by PlagueHO** [comment]
Personally I'd be happy with @ronx 's suggestion as any repos I work on I do this in anyway. However making this a style guideline does introduce a bit of tech debt. However, with this setting enabled in VS Code then this tech debt is trivial to resolve.

My question: is there really people out there not using VS Code for PowerShell DSC development? I have ISE Steroids and it is really fantastic for some specific scenarios, but I've not touched it in months.

**Comment by johlju** [comment]
I all for having this in the guideline, so then contributors not having VS Code (and not having the formatting settings) the code will look the same as those who actually uses the VS Code with the default auto-formatting settings (which would align property value pairs).

This was formatted with VS Code. In this scenario, I think it does not become more readable, but these occasion are so rare I think. Although, the below might be a bug, maybe it should format these when they are so far apart. 

```powershell
$ConfigurationData = @{
    AllNodes = @(
        @{
            NodeName                    = '*'

            <#
                NOTE! THIS IS NOT RECOMMENDED IN PRODUCTION.
                This is added so that AppVeyor automatic tests can pass, otherwise the tests will fail on
                passwords being in plain text and not being encrypted. Because there is not possible to have
                a certificate in AppVeyor to encrypt the passwords we need to add parameter
                'PSDscAllowPlainTextPassword'.
                NOTE! THIS IS NOT RECOMMENDED IN PRODUCTION.
            #>
            PSDscAllowPlainTextPassword = $true

            <#
                Replace with the thumbprint of certificate that are installed on both the target nodes.
                This must be the private certificate of the same public certificate used in the previous
                parameter CertificateFile.
                For this example it is assumed that both machines have the same certificate installed.
            #>
            Thumbprint                  = "E513EEFCB763E6954C52BA66A1A81231BF3F551E"

            <#
                Replace with your own CNO (Cluster Name Object) and IP address.

                Please note that if the CNO is prestaged, then the computer object must be disabled for the
                resource xCluster to be able to create the cluster.
                If the CNO is not prestaged, then the credential used in the xCluster resource must have
                the permission in Active Directory to create the CNO (Cluster Name Object).
            #>
            ClusterName                 = 'Cluster01'
            ClusterIPAddress            = '192.168.100.20/24'
        },
}
```

@PlagueHO There are probably people out that still has not moved over to VS Code. Would be nice to do a survey of that. 😄 But I would guess that most is using VS Code.

**Comment by bgelens** [comment]
For the people not using vs code, we should provide the instructions on how to manually apply the rules using [Invoke-Formatter](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/docs/markdown/Invoke-Formatter.md)

**Comment by PlagueHO** [comment]
@bgelens - I didn't even know about the `Invoke-Formatter` - awesome!

One idea might be to include the `Settings/CodeFormatting.psd1` file in the `DSCResources.Tests` repo and add a helper function into MetaFixers.psm1 that can go through all `PS1` or `PSM1` files in a repo and apply the formatting.

However, I did some experimentation with the 1.15 PSScriptAnalyzer and no matter which rules I'm applying to the PS1/PSM1 files in a repo it never seems to identify any violations or makes any changes:
![image](https://user-images.githubusercontent.com/7589164/27853103-a8ada860-61b5-11e7-9c28-a9b531ccf55f.png)

Have you had any luck using this to fix any formatting problems?

**Comment by bgelens** [comment]
I've used it in build scripts like this:
```powershell
$function = Get-Content -Path $ps1.FullName -Raw
Invoke-Formatter -ScriptDefinition $function -Settings CodeFormattingOTBS |
    Out-File -Encoding utf8 -FilePath "$distFolder\module.psm1" -Append
```
So you need to pass the content and when it is invoked,  it outputs the corrected code and you need to handle that again.

**Comment by johlju** [comment]
Regarding the `Invoke-Formatter` it seems it would need documentation to show users how to use that and how to diff the change? With `Invoke-Formatter` there is not a visibility to what was changed. You need to diff the file with the previous to see what was changed, and it it was changed correctly. This is already present in VSCode, and a maintainer just need the user to do <kbd>alt</kbd>+<kbd>shift</kbd>+<kbd>f</kbd>, and a user can easily check the diff i VSCode before the commit.
If someone wants to add a standard way of using `Invoke-Formatting` together with the documentation then I suggest we open up another issue for that.

Moving back to the main part of this issue, the "Visual Studio Code PowerShell extension formatting settings to DSC Resource template"

Many repos have added the file .vscode/settings.json with this change to be able to get the correct formatting in Visual Studio Code, but there have not yet been any updates in this repo for guidance.
I suggest we add this file .vscode/settings.json (with the correct content above) to the [DscResource.Template](
https://github.com/PowerShell/DscResources/tree/master/DscResource.Template).
